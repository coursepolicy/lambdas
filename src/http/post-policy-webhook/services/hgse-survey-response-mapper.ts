import {
  generativeAiPolicyMapper,
  hgseUseCaseReasonabilityMapper, HgseUseCases,
  MappedSurveyResponse,
  SurveyResponse,
  useCasesMapper,
  Values
} from "../../../shared";
import {toTitleCase} from "../utils/helpers";

const formatUseCases = ({
  QID16_DO,
  QID19_TEXT,
  QID20_TEXT,
  values,
}: {
  QID16_DO?: string[];
  QID19_TEXT?: string;
  QID20_TEXT?: string;
  values: Values;
}): HgseUseCases | undefined => {
  let useCases;

  if (QID16_DO) {
    useCases = QID16_DO.reduce(
      (acc: HgseUseCases, item: string) => {
        let [label, text] = item.split(':');
        label = toTitleCase(label);
        text = text.trim();

        const key = useCasesMapper[label];
        const statusNum = values[key];
        const status = hgseUseCaseReasonabilityMapper[Number(statusNum)];

        const useCaseEntry = {
          label,
          text,
        };

        if (status === 'Not Applicable') {
          return acc;
        }

        if (status === 'Acceptable') {
          acc.acceptable.push(useCaseEntry);
          return acc;
        }

        if (status === 'Not Acceptable') {
          acc.unacceptable.push(useCaseEntry);
          return acc;
        }
        return acc;
      },
      {
        acceptable: [],
        unacceptable: [],
      }
    );

    if (QID19_TEXT) {
      useCases.acceptable.push({
        label: 'Additional examples',
        text: QID19_TEXT,
      });
    }

    if (QID20_TEXT) {
      useCases.unacceptable.push({
        label: 'Additional examples',
        text: QID20_TEXT,
      });
    }
  }

  return useCases;
};

export const hgseSurveyResponseMapper = ({
 values,
 labels,
}: SurveyResponse): MappedSurveyResponse<HgseUseCases> => {
  const {
    endDate,
    QID15,
    QID25_6_TEXT: additionalGuidelines,
    QID19_TEXT,
    QID20_TEXT,
    QID17_TEXT,
    QID26_3_TEXT,
    QID30_TEXT,
    QID4_3,
    QID13_TEXT,
    QID4_4,
    QID4_1,
    QID4_2,
    QID12_1,
    QID12_2,
    QID12_3,
    QID12_4,
  } = values;
  const { QID16_DO, QID26_DO, QID25 } = labels;

  const ethicalGuidelines = additionalGuidelines
    ? QID25.slice(0, QID25.length - 1)
    : undefined; // If QID25_6_TEXT exists, extract QID25
  const generativeAiToolDeclarations = QID26_DO
    ? QID26_DO.slice(0, QID26_DO.length - 1)
    : undefined;
  const useCases = formatUseCases({
    QID19_TEXT,
    QID20_TEXT,
    QID16_DO,
    values,
  });

  const base = {
    id: QID13_TEXT,
    courseNumber: QID4_3,
    courseTitle: QID4_4,
    instructor: QID4_1,
    email: QID4_2,
    overallPolicy: generativeAiPolicyMapper[QID15],
    additionalGuidelines,
    ethicalGuidelines,
    useCases,
    specificPoliciesForAssignments: QID17_TEXT,
    generativeAiToolDeclarations,
    additionalGenerativeAiToolsDeclarations: QID26_3_TEXT,
    additionalNotes: QID30_TEXT,
    createdAt: endDate,
  };

  // save button -> ulid or link sent to email

  if (generativeAiPolicyMapper[QID15] === 'No restrictions') {
    return {
      ...base,
      overallPolicyText: `This course imposes no restrictions on the use of generative AI, recognizing each student's unique learning methods. This flexibility empowers you to tailor your educational journey to your needs. However, while exploring these technological options, adherence to the department or school's academic integrity policies is essential. This ensures that while maximizing your learning potential, you also uphold our community's high standards of academic ethics.`,
      additionalPolicyText:
        "This policy document aims to provide clarity and transparency for the use of generative AI in our course. However, it's paramount to remember that students are also expected to adhere to all other policies specified in the course syllabus and those established by the school administration.",
    };
  }
  if (generativeAiPolicyMapper[QID15] === 'Allowed under conditions') {
    return {
      ...base,
      campusWidePolicy: QID12_1,
      departmentWidePolicy: QID12_2,
      academicIntegrityPolicy: QID12_3,
      otherPolicies: QID12_4,
      overallPolicyText: `We recognize the potential benefits of incorporating generative AI in the learning process. As such, we embrace the use of generative AI tools by our students. In this policy, we employ a "reasonable/not reasonable" system rather than a strict "allowed/not allowed" one (inspired by CS50 at Harvard). This approach fosters proactive thinking among students by encouraging them to understand context, evaluate implications, and make thoughtful decisions. `,
      additionalPolicyText:
        "This policy document is designed to clarify and transparently outline the use of generative AI in our course. It is crucial for students to also comply with all additional university-wide and school-wide policies listed below.",
    };
  }
  if (generativeAiPolicyMapper[QID15] === 'Strictly prohibited') {
    return {
      ...base,
      campusWidePolicy: QID12_1,
      academicIntegrityPolicy: QID12_3,
      overallPolicyText: `The use of generative AI is strictly prohibited in this course to optimize students' learning outcomes. This policy is instituted to inspire comprehensive engagement with the course content and foster a deep understanding of the subject matter. It provides an avenue for students to articulate their ideas, form personal connections with the material, and bolster their academic development.`,
      additionalPolicyText:
        "This policy document aims to provide clarity and transparency for the use of generative AI in our course. However, it's paramount to remember that students are also expected to adhere to all other policies specified in the course syllabus and those established by the school administration. The following represents a non-exhaustive list of institution-wide policies which all students must observe, some of which may touch on the use of generative AI. These policies are subject to modification at any point in time. It's incumbent upon the students to keep themselves updated and well-informed about these policies. ",
    };
  }
  throw new Error('Invalid policy');
};
