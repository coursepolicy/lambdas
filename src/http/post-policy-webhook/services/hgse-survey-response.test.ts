import {hgseSurveyResponseMapper} from "./hgse-survey-response-mapper"; // Update with the correct path
import {
  DisplayedValues,
  Labels,
  SurveyResponse,
  Values,
} from '../../../shared';

describe('hgseSurveyResponseMapper', () => {
  it('should match the snapshot for a given response', () => {
    const mockLabels: Labels = {
      QID8: 'SomeQID8',
      QID15: 'Allowed under conditions',
      QID22: 'SomeQID22',
      QID24: 'SomeQID24',
      QID25: ['Guideline1', 'Guideline2', 'Guideline3'],
      QID26: 'SomeQID26',
      QID28: 'SomeQID28',
      status: 'SomeStatus',
      QID4_DO: ['Action1', 'Action2'],
      QID8_DO: ['Response1', 'Response2'],
      QID12_DO: ['ResponseA', 'ResponseB'],
      QID15_DO: ['Acceptable', 'Not NotAcceptable'],
      QID16_DO: ['grammar check: Test', 'concept learning: Test'],
      QID22_DO: ['Option1', 'Option2'],
      QID24_DO: ['OptionA', 'OptionB'],
      QID25_DO: ['GuidelineA', 'GuidelineB'],
      QID26_DO: ['ToolDeclaration1', 'ToolDeclaration2'],
      QID28_DO: ['ValueA', 'ValueB'],
      finished: 'Completed',
    };

    const mockValues: Values = {
      QID8: 1,
      QID15: 2,
      QID22: 3,
      QID24: 4,
      QID25: ['GuidelineValue1', 'GuidelineValue2'],
      QID26: 5,
      QID28: 6,
      QID4_1: 'InstructorXYZ',
      QID4_2: 'email@example.com',
      QID4_3: 'CourseNumber123',
      QID4_4: 'CourseTitleABC',
      status: 7,
      QID12_1: 'SomeCampusPolicy',
      QID12_2: 'SomeDepartmentPolicy',
      QID12_3: 'SomeIntegrityPolicy',
      QID12_4: 'SomeOtherPolicies',
      QID16_1: 1,
      QID16_2: 2,
      QID4_DO: ['ActionValue1', 'ActionValue2'],
      QID8_DO: ['ResponseValue1', 'ResponseValue2'],
      endDate: '2023-09-08',
      QID12_DO: ['ResponseValueA', 'ResponseValueB'],
      QID15_DO: ['ReasonableValue', 'NotReasonableValue'],
      QID16_DO: ['Label:AcceptableValue', 'Label:NotAcceptableValue'],
      QID22_DO: ['OptionValue1', 'OptionValue2'],
      QID24_DO: ['OptionValueA', 'OptionValueB'],
      QID25_DO: ['GuidelineValueA', 'GuidelineValueB'],
      QID26_DO: ['ToolDeclarationValue1', 'ToolDeclarationValue2'],
      QID28_DO: ['ValueA', 'ValueB'],
      duration: 120,
      finished: 100,
      progress: 80,
      ipAddress: '192.168.1.1',
      startDate: '2023-09-01',
      QID13_TEXT: 'SomeID',
      QID17_TEXT: 'SomeSpecificPolicies',
      QID19_TEXT: 'SomeQID19Text',
      QID20_TEXT: 'SomeQID20Text',
      QID25_6_TEXT: 'SomeAdditionalGuidelines',
      QID26_3_TEXT: ['SomeGenerativeAiTools1', 'SomeGenerativeAiTools2'],
      QID30_TEXT: 'SomeNotes',
      recordedDate: '2023-09-07',
      userLanguage: 'en-US',
      Q_RecaptchaScore: 0.9,
      locationLatitude: '40.730610',
      locationLongitude: '-73.935242',
      distributionChannel: 'Web',
    };

    const mockResponse: SurveyResponse = {
      responseId: 'SomeResponseId',
      displayedFields: [],
      displayedValues: {} as DisplayedValues,
      labels: mockLabels,
      values: mockValues,
    };

    const result = hgseSurveyResponseMapper(mockResponse);

    expect(result).toMatchSnapshot();
  });
});
