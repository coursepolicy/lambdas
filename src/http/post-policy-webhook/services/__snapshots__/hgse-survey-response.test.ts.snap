// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hgseSurveyResponseMapper should match the snapshot for a given response 1`] = `
{
  "academicIntegrityPolicy": "SomeIntegrityPolicy",
  "additionalGenerativeAiToolsDeclarations": [
    "SomeGenerativeAiTools1",
    "SomeGenerativeAiTools2",
  ],
  "additionalGuidelines": "SomeAdditionalGuidelines",
  "additionalNotes": "SomeNotes",
  "additionalPolicyText": "This policy document is designed to clarify and transparently outline the use of generative AI in our course. It is crucial for students to also comply with all additional university-wide and school-wide policies listed below.",
  "campusWidePolicy": "SomeCampusPolicy",
  "courseNumber": "CourseNumber123",
  "courseTitle": "CourseTitleABC",
  "createdAt": "2023-09-08",
  "departmentWidePolicy": "SomeDepartmentPolicy",
  "email": "email@example.com",
  "ethicalGuidelines": [
    "Guideline1",
    "Guideline2",
  ],
  "generativeAiToolDeclarations": [
    "ToolDeclaration1",
  ],
  "id": "SomeID",
  "instructor": "InstructorXYZ",
  "otherPolicies": "SomeOtherPolicies",
  "overallPolicy": "Allowed under conditions",
  "overallPolicyText": "We recognize the potential benefits of incorporating generative AI in the learning process. As such, we embrace the use of generative AI tools by our students. In this policy, we employ a "reasonable/not reasonable" system rather than a strict "allowed/not allowed" one (inspired by CS50 at Harvard). This approach fosters proactive thinking among students by encouraging them to understand context, evaluate implications, and make thoughtful decisions. ",
  "specificPoliciesForAssignments": "SomeSpecificPolicies",
  "useCases": {
    "acceptable": [
      {
        "label": "Grammar Check",
        "text": "Test",
      },
      {
        "label": "Additional examples",
        "text": "SomeQID19Text",
      },
    ],
    "unacceptable": [
      {
        "label": "Concept Learning",
        "text": "Test",
      },
      {
        "label": "Additional examples",
        "text": "SomeQID20Text",
      },
    ],
  },
}
`;
